server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SubAddressingTest.java

```diff
@@ -0,0 +1,213 @@
+        recipientUsername = Username.of("recipient");
+        recipientSession = mailboxManager.createSystemSession(recipientUsername);
+        targetMailboxId = mailboxManager.createMailbox(
+            MailboxPath.forUser(recipientUsername, TARGET), recipientSession).get();
+    }
+
+    @Test
+    void shouldNotAddStorageDirectiveWhenTargetMailboxDoesNotExist() throws Exception {
+        Mail mail = mailBuilder(UNEXISTING_TARGET).sender(SENDER1).build();
+        testee.service(mail);
+
+        AttributeName recipient = AttributeName.of("DeliveryPaths_recipient@localhost");
+        assertThat(mail.attributes().map(this::unbox))
+            .doesNotContain(Pair.of(recipient, UNEXISTING_TARGET));
+    }
+
+    @Test
+    void shouldNotAddStorageDirectiveWhenNobodyHasRight() throws Exception {
+        givePostRightForKey(MailboxACL.ANYBODY_NEGATIVE_KEY);
+
+        Mail mail = mailBuilder(TARGET).sender(SENDER1).build();
+        testee.service(mail);
+
+        AttributeName recipient = AttributeName.of("DeliveryPaths_recipient@localhost");
+        assertThat(mail.attributes().map(this::unbox))
+            .doesNotContain(Pair.of(recipient, TARGET));
+    }
+
+
+    @Test
+    void shouldAddStorageDirectiveWhenAnybodyHasRight() throws Exception {
+        givePostRightForKey(MailboxACL.ANYBODY_KEY);
+
+        Mail mail = mailBuilder(TARGET).sender(SENDER1).build();
+        testee.service(mail);
+
+        AttributeName recipient = AttributeName.of("DeliveryPaths_recipient@localhost");
+        assertThat(mail.attributes().map(this::unbox))
+            .containsOnly(Pair.of(recipient, TARGET));
+    }
+
+    @Test
+    void shouldAddStorageDirectiveWhenSenderIsWhiteListed() throws Exception {
+        // whitelist sender 1 and send from sender 1
+        givePostRightForKey(MailboxACL.EntryKey.createUserEntryKey(sender1Username, false));
+
+        Mail mail = mailBuilder(TARGET).sender(SENDER1).build();
+        testee.service(mail);
+
+        AttributeName recipient = AttributeName.of("DeliveryPaths_recipient@localhost");
+        assertThat(mail.attributes().map(this::unbox))
+            .containsOnly(Pair.of(recipient, TARGET));
+    }
+
+    @Test
+    void shouldNotAddStorageDirectiveWhenSenderIsNotWhiteListed() throws Exception {
+        // whitelist sender 1 and send from sender 2
+        givePostRightForKey(MailboxACL.EntryKey.createUserEntryKey(sender1Username, false));
+
+        Mail mail = mailBuilder(TARGET).sender(SENDER2).build();
+        testee.service(mail);
+
+        AttributeName recipient = AttributeName.of("DeliveryPaths_recipient@localhost");
+        assertThat(mail.attributes().map(this::unbox))
+            .doesNotContain(Pair.of(recipient, TARGET));
+    }
+
+    @Test
+    void shouldNotAddStorageDirectiveWhenSenderIsBlackListed() throws Exception {
+        // blacklist sender 1 and send from sender 1
+        givePostRightForKey(MailboxACL.EntryKey.createUserEntryKey(sender1Username, true));
```

Comment:
true is a magic constant. Can't guess easily what it means. Extracting it as a constant could allow naming and help express the intent.

Also we could improve EntryKey api:

```
// API stability
MailboxACL.EntryKey.createUserEntryKey(sender1Username, false|false)

MailboxACL.EntryKey.createUserEntryKey(sender1Username)

MailboxACL.EntryKey.createNegativeUserEntryKey(sender1Username)
```